ghproxy/ghmetrics/ghmetrics.go

```diff
@@ -0,0 +1,61 @@
+package ghmetrics
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+// ghTokenUsageGaugeVec provides the 'github_token_usage' gauge that
+// enables keeping track of GitHub calls and quotas.
+var ghTokenUsageGaugeVec = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "github_token_usage",
+		Help: "How many GitHub token requets have been used up.",
+	},
+	[]string{"remaining", "until_reset"},
+)
+
+// ghRequestsGauge provides the 'github_requests' gauge that keeps track
+// of the number of GitHub requests by API path.
+var ghRequestsGauge = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "github_requests",
+		Help: "GitHub requests by API path.",
+	},
+	[]string{"path", "status", "duration"},
+)
+
+func init() {
+	prometheus.MustRegister(ghTokenUsageGaugeVec)
+	prometheus.MustRegister(ghRequestsGauge)
+}
+
+// CollectGithubTokenMetrics publishes the rate limits of the github api to
+// `github_token_usage` on prometheus.
+func CollectGithubTokenMetrics(headers http.Header, now time.Time) {
```

Comment:
I don't think this is threadsafe, we should have a serialized critical section around the metric set and conditionally update based on newer `until_reset` fields to handle out-of-order updates

Reply:
How would I go about this? Put a mutex on the metric write, is that what you mean?

Reply:
Yeah, a lock that's hidden in this package and used to update the gauges would work -- also with the `until_reset` check on top